backend/src/dao/ShoutoutsDao.ts

```diff
@@ -1,86 +1,59 @@
   }
 
-  static async getShoutouts(email: string, type: 'given' | 'received'): Promise<Shoutout[]> {
+  async getAllShoutouts(): Promise<Shoutout[]> {
+    return this.getAllDocuments();
+  }
+
+  async getShoutouts(email: string, type: 'given' | 'received'): Promise<Shoutout[]> {
     const givenOrReceived = type === 'given' ? 'giver' : 'receiver';
-    const shoutoutRefs = await shoutoutCollection
+    const shoutoutRefs = await this.collection
       .where(givenOrReceived, '==', memberCollection.doc(email))
       .get();
     return Promise.all(
-      shoutoutRefs.docs.map(async (shoutoutRef) => {
-        const { giver, receiver, message, isAnon, timestamp, hidden, uuid } = shoutoutRef.data();
-        return {
-          giver: await getMemberFromDocumentReference(giver),
-          receiver,
-          message,
-          isAnon,
-          timestamp,
-          hidden,
-          uuid
-        };
-      })
+      shoutoutRefs.docs.map(async (shoutoutRef) => materializeShoutout(shoutoutRef.data()))
     );
   }
 
-  static async getShoutout(uuid: string): Promise<Shoutout | undefined> {
-    const doc = await shoutoutCollection.doc(uuid).get();
-    const dbShoutout = doc.data();
-    if (!dbShoutout) return undefined;
-    return {
-      ...dbShoutout,
-      giver: await getMemberFromDocumentReference(dbShoutout.giver)
-    };
+  async getShoutout(uuid: string): Promise<Shoutout | null> {
+    return this.getDocument(uuid);
   }
 
-  static async setShoutout(shoutout: {
-    giver: IdolMember;
-    receiver: string;
-    message: string;
-    isAnon: boolean;
-    timestamp: number;
-    hidden: boolean;
-    uuid: string;
-  }): Promise<Shoutout> {
-    const shoutoutRef: DBShoutout = {
+  async createShoutout(shoutout: Shoutout): Promise<Shoutout> {
+    const shoutoutWithUUID = {
       ...shoutout,
-      giver: memberCollection.doc(shoutout.giver.email),
       uuid: shoutout.uuid ? shoutout.uuid : uuidv4()
     };
-    await shoutoutCollection.doc(shoutoutRef.uuid).set(shoutoutRef);
-    return shoutout;
+    return this.createDocument(shoutoutWithUUID.uuid, shoutoutWithUUID);
   }
 
-  static async updateShoutout(shoutout: Shoutout): Promise<Shoutout> {
-    const shoutoutDoc = shoutoutCollection.doc(shoutout.uuid);
-    const ref = await shoutoutDoc.get();
-    if (!ref.exists) throw new NotFoundError(`No shoutout '${shoutout.uuid}' exists.`);
-    const shoutoutRef: DBShoutout = {
-      ...shoutout,
-      giver: memberCollection.doc(shoutout.giver.email)
-    };
-    await shoutoutCollection.doc(shoutout.uuid).update(shoutoutRef);
-    return shoutout;
+  async updateShoutout(shoutout: Shoutout): Promise<Shoutout> {
+    return this.updateDocument(shoutout.uuid, shoutout);
   }
 
-  static async deleteShoutout(uuid: string): Promise<void> {
-    await shoutoutCollection.doc(uuid).delete();
+  async deleteShoutout(uuid: string): Promise<void> {
+    await this.deleteDocument(uuid);
```

Comment:
I see the conversation you and @andrew032011 had about these naming conventions. I also see you made the design decision to make the crud methods from the base dao protected, but I feel like these public wrapper methods can be redundant sometimes. I think the DAO layer as a whole would be easier to use and extend if naming was uniform for its basic operations `getDocument`, `deleteDocument`, etc. If logic needs to be slightly different before the call to the inherited method, there could be an override instead of a wrapper.

Reply:
Somewhat related to what's below, but I think the dao class for each of the types still needs to implement its own create and update methods bc we're using generics and TypeScript isn't able to infer what the document id should just from generic types. We could definitely just remove the `delete` and `get` wrappers for each of the classes since they're really just wrappers, but then we're left with a class that sometimes uses `Document` and sometimes uses the type that we're acting on.